src/Swarm/Game/Scenario/Topography/Navigation/Portal.hs

```diff
@@ -0,0 +1,120 @@
+{-# LANGUAGE OverloadedStrings #-}
+
+-- |
+-- SPDX-License-Identifier: BSD-3-Clause
+module Swarm.Game.Scenario.Topography.Navigation.Portal where
+
+import Control.Monad (forM, forM_, unless)
+import Data.Aeson (FromJSON)
+import Data.Int (Int32)
+import Data.List (intercalate)
+import Data.List.NonEmpty (NonEmpty ((:|)))
+import Data.List.NonEmpty qualified as NE
+import Data.Map qualified as M
+import Data.Maybe (listToMaybe)
+import Data.Text (Text)
+import Data.Text qualified as T
+import GHC.Generics (Generic)
+import Linear (V2)
+import Swarm.Game.Location
+import Swarm.Game.Scenario.Topography.Navigation.Waypoint
+import Swarm.Util (binTuples, quote)
+
+-- | Note: The primary overworld shall use
+-- the reserved name \"root\".
+newtype SubworldName = SubworldName Text
+  deriving (Show, Eq, Ord, Generic, FromJSON)
+
+data Navigation = Navigation
+  { waypoints :: M.Map WaypointName (NonEmpty Location)
+  -- ^ Note that waypoints defined at the "root" level are still relative to
+  -- the top-left corner of the map rectangle; they are not in absolute world
+  -- coordinates (as with applying the "ul" offset).
+  , portals :: M.Map Location Location
+  }
+  deriving (Eq, Show)
+
+data PortalExit = PortalExit
+  { exit :: WaypointName
+  , subworldName :: Maybe SubworldName
+  -- ^ Note: 'Nothing' indicates that references a waypoint within the same subworld.
+  }
+  deriving (Show, Eq, Generic, FromJSON)
+
+data Portal = Portal
+  { entrance :: WaypointName
+  , exitInfo :: PortalExit
+  }
+  deriving (Show, Eq, Generic, FromJSON)
+
+failUponDuplication ::
+  (MonadFail m, Show a, Show b) =>
+  String ->
+  M.Map a (NonEmpty b) ->
+  m ()
+failUponDuplication message binnedMap =
+  forM_ (listToMaybe $ M.toList duplicated) $ \(pIn, pOuts) ->
+    fail $
+      unwords
+        [ "Waypoint"
+        , show pIn
+        , message
+        , intercalate ", " $ map show $ NE.toList pOuts
+        ]
+ where
+  duplicated = M.filter ((> 1) . NE.length) binnedMap
+
+-- | Enforces the following constraints:
+-- * portals can have multiple entrances but only a single exit
+-- * no two portals share the same entrance location
+-- * global waypoint uniqueness when the "unique" flag is specified
+validateNavigation ::
+  (MonadFail m, Traversable t) =>
+  V2 Int32 ->
```

Comment:
Why is this a `V2 Int32` as opposed to a `Location`?

Reply:
I see how that is confusing; I [use it as a vector to offset a location](https://github.com/swarm-game/swarm/pull/1356/files#r1260462962), but really I should pass it into this function as a `Location` and defer the conversion.  Will follow up with fix.